### ISSUE TITLE
Not compatible with Python 3

### ISSUE THREAD

**Issue by jonathonf** [ToxicFrog/Ligaturizer#6]
It looks like the script isn't Python 3-compatible:

```
$ fontforge -script /build/sources/ligaturize.py
Copyright (c) 2000-2014 by George Williams. See AUTHORS for Contributors.
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
 with many parts BSD <http://fontforge.org/license.html>. Please read LICENSE.
 Based on sources from 23:10 UTC  5-Aug-2017-ML-D.
 Based on source from git with hash: 
  File "/build/sources/ligaturize.py", line 316
    print "Generated ligaturized font %s in %s" % (output_font['fullname'], output_full_path)
                                              ^
SyntaxError: invalid syntax
```

It doesn't look like there are many changes needed, though:

```
$ 2to3 /build/sources/ligaturize.py
RefactoringTool: Skipping optional fixer: buffer
RefactoringTool: Skipping optional fixer: idioms
RefactoringTool: Skipping optional fixer: set_literal
RefactoringTool: Skipping optional fixer: ws_comma
RefactoringTool: Refactored /build/sources/ligaturize.py
--- /build/sources/ligaturize.py	(original)
+++ /build/sources/ligaturize.py	(refactored)
@@ -9,7 +9,7 @@
 COPYRIGHT = '\nProgramming ligatures added by Ilya Skriblovsky from FiraCode\nFiraCode Copyright (c) 2015 by Nikita Prokopov'
 
 def get_input_fontname():
-    return raw_input('Enter the source font filename (including extension): ')
+    return input('Enter the source font filename (including extension): ')
 
 def get_input_path(input_fontname):
     return SOURCE_FONT_DIR + "/" + input_fontname
@@ -28,7 +28,7 @@
     return no_variant[:-1] if (no_variant.endswith(" ") or no_variant.endswith("-")) else no_variant
 
 def get_output_fontname(input_name):
-    new_fontname = raw_input('Enter a name for your ligaturized font -- or press ENTER to use the same name: ')
+    new_fontname = input('Enter a name for your ligaturized font -- or press ENTER to use the same name: ')
     if new_fontname == "":
         new_fontname = input_name
     return name_without_width_variant(name_without_file_extension(new_fontname))
@@ -298,7 +298,7 @@
     try:
         creator.add_ligature(lig_spec['chars'], lig_spec['firacode_ligature_name'])
     except Exception as e:
-        print('Exception while adding ligature: {}'.format(lig_spec))
+        print(('Exception while adding ligature: {}'.format(lig_spec)))
         raise
 
 change_font_names(font, output_font['fontname'],
@@ -313,4 +313,4 @@
 output_full_path = OUTPUT_FONT_DIR + "/" + output_name
 font.generate(output_name)
 os.rename(output_name, output_full_path)
-print "Generated ligaturized font %s in %s" % (output_font['fullname'], output_full_path)
+print("Generated ligaturized font %s in %s" % (output_font['fullname'], output_full_path))
RefactoringTool: Files that need to be modified:
RefactoringTool: /build/sources/ligaturize.py
```

**Comment by ToxicFrog** [comment]
Unfortunately, that's not enough on its own; when run under python3, even with those changes, it barfs:

```
    self.font[char].addPosSub(lookup_sub_name(i), cr_name(i))
TypeError: expected bytes, str found
```

Unfortunately, I can't figure out what it's actually objecting to here. Changing the type of the first or second argument results in it saying `argument N should be str, not bytes`. Changing the type of the index causes it to complain that the font can only be indexed by strings or integers. So I have no idea what there is a string when it should be bytes.

**Comment by ToxicFrog** [comment]
There's also issues calling `font.addLookup`, and looking at the Fontforge github page, it looks like this is a known issue with Fontforge in combination with Python 3.5+; it gets confused about functions taking strings and wants them to be both strings and bytes. This may have to wait until [fontforge#3057](https://github.com/fontforge/fontforge/issues/3057) is fixed.

**Comment by ToxicFrog** [comment]
Fontforge links directory to libpython, as I understand it; you need to specify py2 or py3 at the time you build fontforge. It looks like the PKGBUILD there just depends on `python`, which in Arch is Python 3.x.

If you edit the PKGBUILD to depend on `python2` and build fontforge using that you should get a build that uses py2.

**Comment by Kidlike** [comment]
So currently the only solution to use this project in _up-to-date_ systems is to build `fontforge` from sources for python 2 ? :disappointed:

**Comment by ToxicFrog** [comment]
Even "up to date" distros _should_ still be shipping Fontforge linked against Python2, because Py3 Fontforge is broken -- it's not just this script that's affected. I know Void Linux and NixOS are still building it with Py2, and I know that OpenSUSE is (unfortunately) using Py3; not sure about other distros.

If you're more willing to install new package managers than to build stuff from source by hand, you could install Nix and then `nix-shell -p fontforge` to get a shell where the py2 version is in `$PATH`.

You could also file a bug against your distro's packaging of it, referencing fontforge#3057.

**Comment by ToxicFrog** [comment]
This may have been fixed by [fontforge#3409](https://github.com/fontforge/fontforge/pull/3409).

**Comment by rojiani** [comment]
`addLookup` seems to be fixed (fontforge#3057), and fontforge#3409 is merged.
Submitted a PR for moving to Python 3: https://github.com/ToxicFrog/Ligaturizer/pull/68

**Comment by phaux** [comment]
This worked for me:

```sh
git clone https://github.com/ToxicFrog/Ligaturizer.git --recursive
cd Ligaturizer
podman run -it --rm -v "$PWD:/app" -w "/app" python:3 bash
apt-get update
apt-get install python-fontforge fontforge
make
```

Docker should work too

**Comment by rojiani** [comment]
Since I can't merge without approval, I've created https://github.com/rojiani/Ligaturizer-2.0 for now.

**Comment by ToxicFrog** [comment]
Merged daa4dc8baffeefcb27c4ffd30ea52797ead8d123. Sorry it took so long.
